<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-del-projecto-team-27/blob/main/Actividad1_Semana3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Parte 1: Fundamentos de bases de datos**

Fundamentos de bases de datos y para ciencia de datos.
Las bases de datos es un punto muy importante en la ciencia de datos, ya que es el punto de partida para que podamos hacer un análisis y posteriormente ejecutar los modelos de predicción que se puedan usar. Para esto existen diferentes bases de datos una de las primordiales que se debe tener conocimiento ya que en la mayoría de las empresas son usadas son “Relational Database Management Systems”. Sin embargo, las herramientas que son usadas para este tipo de base de datos son solo para los datos estructurados, los datos no estructurados están enfocados en otros sistemas y obtener información puede ser un poco más complicado.
En el data set que tengamos por analizar es importante corregir entradas erróneas, así como encontrar partes incompletas o que sean irrelevantes para el análisis, para eso es necesario modificar o eliminar todos esos datos previamente a nuestro modelo. Unos de los pasos a identificar en este proceso son, tipos de datos faltantes, técnicas para manejar los datos faltantes e Identificación de valores atípicos y datos ruidosos.


Fundamentos de almacenes de datos (Data Warehouse) para ciencia de datos.

Un “Data warehouse” es un repositorio de información que puede ser analizado para hacer decisiones más informadas. Los datos corren en un “data warehouse” mediante sistemas que ayuda y dan soporte para actividades de “Business Intelligence”.  Centralizan y consolidan grandes cantidades de datos de diferentes fuentes y sus capacidades analíticas permiten derivar conocimiento valioso mediante el uso de ciencia de datos.

Referencias:


---


Kane, F. (2017). Hands-on data science and python machine learning. Packt Publishing Ltd. 
SQL for Data Scientists  A Beginner's Guide for Building Datasets for Analysis, Renee M. P. Teate
Amazon Web Services. (s/f). Data Warehouse Concepts. Amazon Web Services. https://aws.amazon.com/data-warehouse/



In [121]:
import pandas as pd
import numpy as np

In [122]:
#descargar la base de datos usando pandas
df = pd.read_csv("https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv")
df

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [123]:
df.to_csv("data.csv") #exportar el df para analizar el csv en Excel

##**Nomenclatura de todos los valores**

'''X1= se refiere a la cantidad del crédito
X2= Género
X3= Nivel de educación
X4= Estado civil
X5= Edad
X6= Historial crediticio septiembre
X7= Histc Agosto
X8= Histc Julio
X9= Histc Junio
X10= Histc Mayo
X11 = Histc Abril
X12= Monto de billing Sep
X13= Montbill Ago
X14= Montbill Jul
X15 = Montbill Jun
X16= Montbill Mayo
X17= Montbill Abril
X18= Monto pago anterior Sep
X19 = Montpag Ago
X20= Montpag Julio
X21= Montpag Junio
X22= Montpag Mayo
X23= Montpag Abril'''




In [124]:
#se copia la base de datos para no modificar los datos originales

new_df=df.copy()
new_df

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [125]:
#se renombran las columnas para identificar
new_df=new_df.rename(columns={"X1":"Monto","X2":"Género","X3":"Educ","X4":"Est_Civ","X5":"Edad","X6":"Histcr_Sep","X7":"Histcr_Ago","X8":"Histcr_Jul","X9":"Histcr_Jun","X10":"Histcr_May","X11":"Histcr_Abr","X12":"Mont_Sep","X13":"Mont_Ago","X14":"Mont_Jul","X15":"Mont_Jun","X16":"Mont_May","X17":"Mont_Abr","X18":"Montpago_Sep","X19":"Montpago_Ago","X20":"Montpago_Jul","X21":"Montpago_Jun","X22":"Montpago_May","X23":"Montpago_Abr"})

In [126]:
new_df

ID   Monto  Género  Educ  Est_Civ  Edad  Histcr_Sep  Histcr_Ago  \
0          1   20000     2.0   2.0      1.0  24.0         2.0         2.0   
1          2  120000     2.0   2.0      2.0  26.0        -1.0         2.0   
2          3   90000     2.0   2.0      2.0  34.0         0.0         0.0   
3          4   50000     2.0   2.0      1.0  37.0         0.0         0.0   
4          5   50000     1.0   2.0      1.0  57.0        -1.0         0.0   
...      ...     ...     ...   ...      ...   ...         ...         ...   
29995  29996  220000     1.0   3.0      1.0  39.0         0.0         0.0   
29996  29997  150000     1.0   3.0      2.0  43.0        -1.0        -1.0   
29997  29998   30000     1.0   2.0      2.0  37.0         4.0         3.0   
29998  29999   80000     1.0   3.0      1.0  41.0         1.0        -1.0   
29999  30000   50000     1.0   2.0      1.0  46.0         0.0         0.0   

       Histcr_Jul  Histcr_Jun  ...  Mont_Jun  Mont_May  Mont_Abr  \
0            -1.0        -1.0  ...       0.0       0.0       0.0   
1             0.0         0.0  ...    3272.0    3455.0    3261.0   
2             0.0         0.0  ...   14331.0   14948.0   15549.0   
3             0.0         0.0  ...   28314.0   28959.0   29547.0   
4            -1.0         0.0  ...   20940.0   19146.0   19131.0   
...           ...         ...  ...       ...       ...       ...   
29995         0.0         0.0  ...   88004.0   31237.0   15980.0   
29996        -1.0        -1.0  ...    8979.0    5190.0       0.0   
29997         2.0        -1.0  ...   20878.0   20582.0   19357.0   
29998         0.0         0.0  ...   52774.0   11855.0   48944.0   
29999         0.0         0.0  ...   36535.0   32428.0   15313.0   

       Montpago_Sep  Montpago_Ago  Montpago_Jul  Montpago_Jun  Montpago_May  \
0               0.0         689.0           0.0           0.0           0.0   
1               0.0        1000.0        1000.0        1000.0           0.0   
2            1518.0        1500.0        1000.0        1000.0        1000.0   
3            2000.0        2019.0        1200.0        1100.0        1069.0   
4            2000.0       36681.0       10000.0        9000.0         689.0   
...             ...           ...           ...           ...           ...   
29995        8500.0       20000.0        5003.0        3047.0        5000.0   
29996        1837.0        3526.0        8998.0         129.0           0.0   
29997           0.0           0.0       22000.0        4200.0        2000.0   
29998       85900.0        3409.0        1178.0        1926.0       52964.0   
29999        2078.0        1800.0        1430.0        1000.0        1000.0   

       Montpago_Abr    Y  
0               0.0  1.0  
1            2000.0  1.0  
2            5000.0  0.0  
3            1000.0  0.0  
4             679.0  0.0  
...             ...  ...  
29995        1000.0  0.0  
29996           0.0  0.0  
29997        3100.0  1.0  
29998        1804.0  1.0  
29999        1000.0  1.0  

[30000 rows x 25 columns]

In [127]:
print(new_df.info())



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ID            30000 non-null  int64  
 1   Monto         30000 non-null  int64  
 2   Género        29999 non-null  float64
 3   Educ          29998 non-null  float64
 4   Est_Civ       29998 non-null  float64
 5   Edad          29995 non-null  float64
 6   Histcr_Sep    29997 non-null  float64
 7   Histcr_Ago    29995 non-null  float64
 8   Histcr_Jul    29993 non-null  float64
 9   Histcr_Jun    29991 non-null  float64
 10  Histcr_May    29984 non-null  float64
 11  Histcr_Abr    29986 non-null  float64
 12  Mont_Sep      29989 non-null  float64
 13  Mont_Ago      29989 non-null  float64
 14  Mont_Jul      29987 non-null  float64
 15  Mont_Jun      29985 non-null  float64
 16  Mont_May      29983 non-null  float64
 17  Mont_Abr      29990 non-null  float64
 18  Montpago_Sep  29992 non-nu

In [128]:
new_df.isnull().values.any() #mostrar si tenemos valores nulos

True

In [129]:
new_df.isnull().sum().sort_values(ascending=True) #ordenar los valores nulos en forma ascendente

ID               0
Monto            0
Género           1
Educ             2
Est_Civ          2
Histcr_Sep       3
Y                3
Histcr_Ago       5
Montpago_Abr     5
Edad             5
Histcr_Jul       7
Montpago_Jul     8
Montpago_Sep     8
Histcr_Jun       9
Montpago_Ago     9
Mont_Abr        10
Montpago_May    11
Montpago_Jun    11
Mont_Sep        11
Mont_Ago        11
Mont_Jul        13
Histcr_Abr      14
Mont_Jun        15
Histcr_May      16
Mont_May        17
dtype: int64

In [130]:
new_df.isnull().sum().sum() #mostrar la cantidad total de mis datos nulos

196

In [131]:
print("El total de números de datos nulos sobre el total es de", round(196/30000*100,2),"%")

El total de números de datos nulos sobre el total es de 0.65 %


Solución 1, para descartar las observaciones con valores faltantes

Ya que el total de valores nulos es relativamente muy bajo, se considera que se pueden eliminar estos registros.

In [132]:
new_df.dropna(inplace = True)#elimina todos los registros nulos
print("Total de datos son", " " + str(new_df.shape[0]))

Total de datos son  29958


In [133]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29958 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ID            29958 non-null  int64  
 1   Monto         29958 non-null  int64  
 2   Género        29958 non-null  float64
 3   Educ          29958 non-null  float64
 4   Est_Civ       29958 non-null  float64
 5   Edad          29958 non-null  float64
 6   Histcr_Sep    29958 non-null  float64
 7   Histcr_Ago    29958 non-null  float64
 8   Histcr_Jul    29958 non-null  float64
 9   Histcr_Jun    29958 non-null  float64
 10  Histcr_May    29958 non-null  float64
 11  Histcr_Abr    29958 non-null  float64
 12  Mont_Sep      29958 non-null  float64
 13  Mont_Ago      29958 non-null  float64
 14  Mont_Jul      29958 non-null  float64
 15  Mont_Jun      29958 non-null  float64
 16  Mont_May      29958 non-null  float64
 17  Mont_Abr      29958 non-null  float64
 18  Montpago_Sep  29958 non-nu

In [134]:
new_df.isna().any() #revisamos si hay algún datos nulo, si marca False quiere decir que todos los registros están correctos

ID              False
Monto           False
Género          False
Educ            False
Est_Civ         False
Edad            False
Histcr_Sep      False
Histcr_Ago      False
Histcr_Jul      False
Histcr_Jun      False
Histcr_May      False
Histcr_Abr      False
Mont_Sep        False
Mont_Ago        False
Mont_Jul        False
Mont_Jun        False
Mont_May        False
Mont_Abr        False
Montpago_Sep    False
Montpago_Ago    False
Montpago_Jul    False
Montpago_Jun    False
Montpago_May    False
Montpago_Abr    False
Y               False
dtype: bool

In [135]:
df= pd.read_csv("https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv")
df.isna().any()

ID     False
X1     False
X2      True
X3      True
X4      True
X5      True
X6      True
X7      True
X8      True
X9      True
X10     True
X11     True
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

In [136]:
ndf = df.copy()#regresamos a copiar la base de datos original para intentar otro método
ndf.head()

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15      X16  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0      0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   3455.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0  14948.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0  28959.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0  19146.0   

       X17     X18      X19      X20     X21     X22     X23    Y  
0      0.0     0.0    689.0      0.0     0.0     0.0     0.0  1.0  
1   3261.0     0.0   1000.0   1000.0  1000.0     0.0  2000.0  1.0  
2  15549.0  1518.0   1500.0   1000.0  1000.0  1000.0  5000.0  0.0  
3  29547.0  2000.0   2019.0   1200.0  1100.0  1069.0  1000.0  0.0  
4  19131.0  2000.0  36681.0  10000.0  9000.0   689.0   679.0  0.0  

[5 rows x 25 columns]

In [137]:
ndf.dropna(axis=1,inplace =True) #quitar todas las columnas donde al menos haya un valor nulo
ndf

ID      X1
0          1   20000
1          2  120000
2          3   90000
3          4   50000
4          5   50000
...      ...     ...
29995  29996  220000
29996  29997  150000
29997  29998   30000
29998  29999   80000
29999  30000   50000

[30000 rows x 2 columns]

In [138]:
ndf.isna().any() #analizamos las columnas que se quedaron, vemos que este método no funciona ya que elimina todo

ID    False
X1    False
dtype: bool

In [139]:
ndf.info()#misma conclusión a la anterior, no nos sirve ya que nos elimina gran parte de los datos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   ID      30000 non-null  int64
 1   X1      30000 non-null  int64
dtypes: int64(2)
memory usage: 468.9 KB


In [140]:
ndf= df.copy() #intentamos otro método
ndf

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [141]:
ndf.dropna(how="all", inplace = True)#eliminamos todas las filas donde tengamos valores nulos
ndf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

In [142]:
ndf.isna().any()

ID     False
X1     False
X2      True
X3      True
X4      True
X5      True
X6      True
X7      True
X8      True
X9      True
X10     True
X11     True
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

In [143]:
ndf = df.copy()
ndf

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [144]:
ndf.dropna(thresh=4, inplace = True) # analizamos que se tenga al menos 4 valores nulos para mantenerlo
ndf

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [145]:
ndf.dropna(thresh=5, axis=1,inplace=True)
ndf

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

Solución 2

In [146]:
new_df

ID   Monto  Género  Educ  Est_Civ  Edad  Histcr_Sep  Histcr_Ago  \
0          1   20000     2.0   2.0      1.0  24.0         2.0         2.0   
1          2  120000     2.0   2.0      2.0  26.0        -1.0         2.0   
2          3   90000     2.0   2.0      2.0  34.0         0.0         0.0   
3          4   50000     2.0   2.0      1.0  37.0         0.0         0.0   
4          5   50000     1.0   2.0      1.0  57.0        -1.0         0.0   
...      ...     ...     ...   ...      ...   ...         ...         ...   
29995  29996  220000     1.0   3.0      1.0  39.0         0.0         0.0   
29996  29997  150000     1.0   3.0      2.0  43.0        -1.0        -1.0   
29997  29998   30000     1.0   2.0      2.0  37.0         4.0         3.0   
29998  29999   80000     1.0   3.0      1.0  41.0         1.0        -1.0   
29999  30000   50000     1.0   2.0      1.0  46.0         0.0         0.0   

       Histcr_Jul  Histcr_Jun  ...  Mont_Jun  Mont_May  Mont_Abr  \
0            -1.0        -1.0  ...       0.0       0.0       0.0   
1             0.0         0.0  ...    3272.0    3455.0    3261.0   
2             0.0         0.0  ...   14331.0   14948.0   15549.0   
3             0.0         0.0  ...   28314.0   28959.0   29547.0   
4            -1.0         0.0  ...   20940.0   19146.0   19131.0   
...           ...         ...  ...       ...       ...       ...   
29995         0.0         0.0  ...   88004.0   31237.0   15980.0   
29996        -1.0        -1.0  ...    8979.0    5190.0       0.0   
29997         2.0        -1.0  ...   20878.0   20582.0   19357.0   
29998         0.0         0.0  ...   52774.0   11855.0   48944.0   
29999         0.0         0.0  ...   36535.0   32428.0   15313.0   

       Montpago_Sep  Montpago_Ago  Montpago_Jul  Montpago_Jun  Montpago_May  \
0               0.0         689.0           0.0           0.0           0.0   
1               0.0        1000.0        1000.0        1000.0           0.0   
2            1518.0        1500.0        1000.0        1000.0        1000.0   
3            2000.0        2019.0        1200.0        1100.0        1069.0   
4            2000.0       36681.0       10000.0        9000.0         689.0   
...             ...           ...           ...           ...           ...   
29995        8500.0       20000.0        5003.0        3047.0        5000.0   
29996        1837.0        3526.0        8998.0         129.0           0.0   
29997           0.0           0.0       22000.0        4200.0        2000.0   
29998       85900.0        3409.0        1178.0        1926.0       52964.0   
29999        2078.0        1800.0        1430.0        1000.0        1000.0   

       Montpago_Abr    Y  
0               0.0  1.0  
1            2000.0  1.0  
2            5000.0  0.0  
3            1000.0  0.0  
4             679.0  0.0  
...             ...  ...  
29995        1000.0  0.0  
29996           0.0  0.0  
29997        3100.0  1.0  
29998        1804.0  1.0  
29999        1000.0  1.0  

[29958 rows x 25 columns]

In [147]:
new_df.describe().T #nos devuelve el resumen de la distribución de los datos, el parámetro "T" lo usamos para transponer las filas
#count nos dice la cantidad de datos no nulos
#mean nos dice el promedio
#std la desviación estándar
#min el valor mínimo
#25% el percentil 25
#50% el percentil 50
#75% el percentil 75
#max el valor máximo

count           mean            std       min       25%  \
ID            29958.0   15005.550504    8654.547473       1.0   7516.25   
Monto         29958.0  167555.900928  129737.299088   10000.0  50000.00   
Género        29958.0       1.604012       0.489070       1.0      1.00   
Educ          29958.0       1.853094       0.790471       0.0      1.00   
Est_Civ       29958.0       1.551739       0.521952       0.0      1.00   
Edad          29958.0      35.483443       9.214319      21.0     28.00   
Histcr_Sep    29958.0      -0.017124       1.123989      -2.0     -1.00   
Histcr_Ago    29958.0      -0.134021       1.197171      -2.0     -1.00   
Histcr_Jul    29958.0      -0.166767       1.196026      -2.0     -1.00   
Histcr_Jun    29958.0      -0.221110       1.168419      -2.0     -1.00   
Histcr_May    29958.0      -0.266807       1.132307      -2.0     -1.00   
Histcr_Abr    29958.0      -0.291575       1.149303      -2.0     -1.00   
Mont_Sep      29958.0   51248.119901   73674.949943 -165580.0   3559.25   
Mont_Ago      29958.0   49200.493825   71211.232744  -69777.0   2984.00   
Mont_Jul      29958.0   47032.385273   69385.243340 -157264.0   2664.75   
Mont_Jun      29958.0   43279.335370   64364.684347 -170000.0   2327.50   
Mont_May      29958.0   40328.984578   60826.219326  -81334.0   1762.25   
Mont_Abr      29958.0   38889.925763   59582.883301 -339603.0   1256.00   
Montpago_Sep  29958.0    5664.614460   16568.823518       0.0   1000.00   
Montpago_Ago  29958.0    5925.715468   23055.983864       0.0    835.25   
Montpago_Jul  29958.0    5228.429969   17617.338167       0.0    390.00   
Montpago_Jun  29958.0    4829.873556   15676.205514       0.0    296.25   
Montpago_May  29958.0    4801.481574   15285.552652       0.0    253.25   
Montpago_Abr  29958.0    5220.708025   17788.983767       0.0    118.00   
Y             29958.0       0.221143       0.415023       0.0      0.00   

                   50%        75%        max  
ID             15005.5   22497.75    30000.0  
Monto         140000.0  240000.00  1000000.0  
Género             2.0       2.00        2.0  
Educ               2.0       2.00        6.0  
Est_Civ            2.0       2.00        3.0  
Edad              34.0      41.00       79.0  
Histcr_Sep         0.0       0.00        8.0  
Histcr_Ago         0.0       0.00        8.0  
Histcr_Jul         0.0       0.00        8.0  
Histcr_Jun         0.0       0.00        8.0  
Histcr_May         0.0       0.00        8.0  
Histcr_Abr         0.0       0.00        8.0  
Mont_Sep       22379.0   67190.00   964511.0  
Mont_Ago       21194.5   64027.75   983931.0  
Mont_Jul       20085.5   60183.00  1664089.0  
Mont_Jun       19037.5   54551.25   891586.0  
Mont_May       18104.5   50220.75   927171.0  
Mont_Abr       17067.5   49234.75   961664.0  
Montpago_Sep    2100.0    5007.00   873552.0  
Montpago_Ago    2009.0    5000.00  1684259.0  
Montpago_Jul    1800.0    4511.50   896040.0  
Montpago_Jun    1500.0    4014.75   621000.0  
Montpago_May    1500.0    4040.00   426529.0  
Montpago_Abr    1500.0    4000.00   528666.0  
Y                  0.0       0.00        1.0

In [148]:
em = new_df.Edad.mean()#encontrar la media de Edad
em

35.48344348754924

In [149]:
new_df["Edad"].fillna(value=em, inplace = True) # imputar valores na con la media para edad
new_df

ID   Monto  Género  Educ  Est_Civ  Edad  Histcr_Sep  Histcr_Ago  \
0          1   20000     2.0   2.0      1.0  24.0         2.0         2.0   
1          2  120000     2.0   2.0      2.0  26.0        -1.0         2.0   
2          3   90000     2.0   2.0      2.0  34.0         0.0         0.0   
3          4   50000     2.0   2.0      1.0  37.0         0.0         0.0   
4          5   50000     1.0   2.0      1.0  57.0        -1.0         0.0   
...      ...     ...     ...   ...      ...   ...         ...         ...   
29995  29996  220000     1.0   3.0      1.0  39.0         0.0         0.0   
29996  29997  150000     1.0   3.0      2.0  43.0        -1.0        -1.0   
29997  29998   30000     1.0   2.0      2.0  37.0         4.0         3.0   
29998  29999   80000     1.0   3.0      1.0  41.0         1.0        -1.0   
29999  30000   50000     1.0   2.0      1.0  46.0         0.0         0.0   

       Histcr_Jul  Histcr_Jun  ...  Mont_Jun  Mont_May  Mont_Abr  \
0            -1.0        -1.0  ...       0.0       0.0       0.0   
1             0.0         0.0  ...    3272.0    3455.0    3261.0   
2             0.0         0.0  ...   14331.0   14948.0   15549.0   
3             0.0         0.0  ...   28314.0   28959.0   29547.0   
4            -1.0         0.0  ...   20940.0   19146.0   19131.0   
...           ...         ...  ...       ...       ...       ...   
29995         0.0         0.0  ...   88004.0   31237.0   15980.0   
29996        -1.0        -1.0  ...    8979.0    5190.0       0.0   
29997         2.0        -1.0  ...   20878.0   20582.0   19357.0   
29998         0.0         0.0  ...   52774.0   11855.0   48944.0   
29999         0.0         0.0  ...   36535.0   32428.0   15313.0   

       Montpago_Sep  Montpago_Ago  Montpago_Jul  Montpago_Jun  Montpago_May  \
0               0.0         689.0           0.0           0.0           0.0   
1               0.0        1000.0        1000.0        1000.0           0.0   
2            1518.0        1500.0        1000.0        1000.0        1000.0   
3            2000.0        2019.0        1200.0        1100.0        1069.0   
4            2000.0       36681.0       10000.0        9000.0         689.0   
...             ...           ...           ...           ...           ...   
29995        8500.0       20000.0        5003.0        3047.0        5000.0   
29996        1837.0        3526.0        8998.0         129.0           0.0   
29997           0.0           0.0       22000.0        4200.0        2000.0   
29998       85900.0        3409.0        1178.0        1926.0       52964.0   
29999        2078.0        1800.0        1430.0        1000.0        1000.0   

       Montpago_Abr    Y  
0               0.0  1.0  
1            2000.0  1.0  
2            5000.0  0.0  
3            1000.0  0.0  
4             679.0  0.0  
...             ...  ...  
29995        1000.0  0.0  
29996           0.0  0.0  
29997        3100.0  1.0  
29998        1804.0  1.0  
29999        1000.0  1.0  

[29958 rows x 25 columns]

In [150]:
new_df["Género"].fillna(value = new_df.Género.median(), inplace = True) # imputar valores na con la media para género
new_df

ID   Monto  Género  Educ  Est_Civ  Edad  Histcr_Sep  Histcr_Ago  \
0          1   20000     2.0   2.0      1.0  24.0         2.0         2.0   
1          2  120000     2.0   2.0      2.0  26.0        -1.0         2.0   
2          3   90000     2.0   2.0      2.0  34.0         0.0         0.0   
3          4   50000     2.0   2.0      1.0  37.0         0.0         0.0   
4          5   50000     1.0   2.0      1.0  57.0        -1.0         0.0   
...      ...     ...     ...   ...      ...   ...         ...         ...   
29995  29996  220000     1.0   3.0      1.0  39.0         0.0         0.0   
29996  29997  150000     1.0   3.0      2.0  43.0        -1.0        -1.0   
29997  29998   30000     1.0   2.0      2.0  37.0         4.0         3.0   
29998  29999   80000     1.0   3.0      1.0  41.0         1.0        -1.0   
29999  30000   50000     1.0   2.0      1.0  46.0         0.0         0.0   

       Histcr_Jul  Histcr_Jun  ...  Mont_Jun  Mont_May  Mont_Abr  \
0            -1.0        -1.0  ...       0.0       0.0       0.0   
1             0.0         0.0  ...    3272.0    3455.0    3261.0   
2             0.0         0.0  ...   14331.0   14948.0   15549.0   
3             0.0         0.0  ...   28314.0   28959.0   29547.0   
4            -1.0         0.0  ...   20940.0   19146.0   19131.0   
...           ...         ...  ...       ...       ...       ...   
29995         0.0         0.0  ...   88004.0   31237.0   15980.0   
29996        -1.0        -1.0  ...    8979.0    5190.0       0.0   
29997         2.0        -1.0  ...   20878.0   20582.0   19357.0   
29998         0.0         0.0  ...   52774.0   11855.0   48944.0   
29999         0.0         0.0  ...   36535.0   32428.0   15313.0   

       Montpago_Sep  Montpago_Ago  Montpago_Jul  Montpago_Jun  Montpago_May  \
0               0.0         689.0           0.0           0.0           0.0   
1               0.0        1000.0        1000.0        1000.0           0.0   
2            1518.0        1500.0        1000.0        1000.0        1000.0   
3            2000.0        2019.0        1200.0        1100.0        1069.0   
4            2000.0       36681.0       10000.0        9000.0         689.0   
...             ...           ...           ...           ...           ...   
29995        8500.0       20000.0        5003.0        3047.0        5000.0   
29996        1837.0        3526.0        8998.0         129.0           0.0   
29997           0.0           0.0       22000.0        4200.0        2000.0   
29998       85900.0        3409.0        1178.0        1926.0       52964.0   
29999        2078.0        1800.0        1430.0        1000.0        1000.0   

       Montpago_Abr    Y  
0               0.0  1.0  
1            2000.0  1.0  
2            5000.0  0.0  
3            1000.0  0.0  
4             679.0  0.0  
...             ...  ...  
29995        1000.0  0.0  
29996           0.0  0.0  
29997        3100.0  1.0  
29998        1804.0  1.0  
29999        1000.0  1.0  

[29958 rows x 25 columns]

In [151]:
mm = new_df.Est_Civ.mode()#queremos ver el valor que más se repite en Estado Civil el cual es soltero
mm

0    2.0
dtype: float64

In [152]:
mm[0]

2.0

In [153]:
new_df["Est_Civ"].fillna(value = mm[0], inplace = True) #vamos a llenar los valores na con la moda
new_df

ID   Monto  Género  Educ  Est_Civ  Edad  Histcr_Sep  Histcr_Ago  \
0          1   20000     2.0   2.0      1.0  24.0         2.0         2.0   
1          2  120000     2.0   2.0      2.0  26.0        -1.0         2.0   
2          3   90000     2.0   2.0      2.0  34.0         0.0         0.0   
3          4   50000     2.0   2.0      1.0  37.0         0.0         0.0   
4          5   50000     1.0   2.0      1.0  57.0        -1.0         0.0   
...      ...     ...     ...   ...      ...   ...         ...         ...   
29995  29996  220000     1.0   3.0      1.0  39.0         0.0         0.0   
29996  29997  150000     1.0   3.0      2.0  43.0        -1.0        -1.0   
29997  29998   30000     1.0   2.0      2.0  37.0         4.0         3.0   
29998  29999   80000     1.0   3.0      1.0  41.0         1.0        -1.0   
29999  30000   50000     1.0   2.0      1.0  46.0         0.0         0.0   

       Histcr_Jul  Histcr_Jun  ...  Mont_Jun  Mont_May  Mont_Abr  \
0            -1.0        -1.0  ...       0.0       0.0       0.0   
1             0.0         0.0  ...    3272.0    3455.0    3261.0   
2             0.0         0.0  ...   14331.0   14948.0   15549.0   
3             0.0         0.0  ...   28314.0   28959.0   29547.0   
4            -1.0         0.0  ...   20940.0   19146.0   19131.0   
...           ...         ...  ...       ...       ...       ...   
29995         0.0         0.0  ...   88004.0   31237.0   15980.0   
29996        -1.0        -1.0  ...    8979.0    5190.0       0.0   
29997         2.0        -1.0  ...   20878.0   20582.0   19357.0   
29998         0.0         0.0  ...   52774.0   11855.0   48944.0   
29999         0.0         0.0  ...   36535.0   32428.0   15313.0   

       Montpago_Sep  Montpago_Ago  Montpago_Jul  Montpago_Jun  Montpago_May  \
0               0.0         689.0           0.0           0.0           0.0   
1               0.0        1000.0        1000.0        1000.0           0.0   
2            1518.0        1500.0        1000.0        1000.0        1000.0   
3            2000.0        2019.0        1200.0        1100.0        1069.0   
4            2000.0       36681.0       10000.0        9000.0         689.0   
...             ...           ...           ...           ...           ...   
29995        8500.0       20000.0        5003.0        3047.0        5000.0   
29996        1837.0        3526.0        8998.0         129.0           0.0   
29997           0.0           0.0       22000.0        4200.0        2000.0   
29998       85900.0        3409.0        1178.0        1926.0       52964.0   
29999        2078.0        1800.0        1430.0        1000.0        1000.0   

       Montpago_Abr    Y  
0               0.0  1.0  
1            2000.0  1.0  
2            5000.0  0.0  
3            1000.0  0.0  
4             679.0  0.0  
...             ...  ...  
29995        1000.0  0.0  
29996           0.0  0.0  
29997        3100.0  1.0  
29998        1804.0  1.0  
29999        1000.0  1.0  

[29958 rows x 25 columns]

In [154]:
favs={"Género":new_df.Género.mode()[0], "Edad":new_df["Edad"].mean()} #vamos a determinar los valores de moda y media para las columnas Edad y Género

In [157]:
new_df.Género.fillna(new_df.Género.mode()[0], inplace=True)#Reemplazamos valores na con moda y media respectivamente
new_df.Edad.fillna(new_df.Edad.mean(), inplace=True)
new_df

ID   Monto  Género  Educ  Est_Civ  Edad  Histcr_Sep  Histcr_Ago  \
0          1   20000     2.0   2.0      1.0  24.0         2.0         2.0   
1          2  120000     2.0   2.0      2.0  26.0        -1.0         2.0   
2          3   90000     2.0   2.0      2.0  34.0         0.0         0.0   
3          4   50000     2.0   2.0      1.0  37.0         0.0         0.0   
4          5   50000     1.0   2.0      1.0  57.0        -1.0         0.0   
...      ...     ...     ...   ...      ...   ...         ...         ...   
29995  29996  220000     1.0   3.0      1.0  39.0         0.0         0.0   
29996  29997  150000     1.0   3.0      2.0  43.0        -1.0        -1.0   
29997  29998   30000     1.0   2.0      2.0  37.0         4.0         3.0   
29998  29999   80000     1.0   3.0      1.0  41.0         1.0        -1.0   
29999  30000   50000     1.0   2.0      1.0  46.0         0.0         0.0   

       Histcr_Jul  Histcr_Jun  ...  Mont_Jun  Mont_May  Mont_Abr  \
0            -1.0        -1.0  ...       0.0       0.0       0.0   
1             0.0         0.0  ...    3272.0    3455.0    3261.0   
2             0.0         0.0  ...   14331.0   14948.0   15549.0   
3             0.0         0.0  ...   28314.0   28959.0   29547.0   
4            -1.0         0.0  ...   20940.0   19146.0   19131.0   
...           ...         ...  ...       ...       ...       ...   
29995         0.0         0.0  ...   88004.0   31237.0   15980.0   
29996        -1.0        -1.0  ...    8979.0    5190.0       0.0   
29997         2.0        -1.0  ...   20878.0   20582.0   19357.0   
29998         0.0         0.0  ...   52774.0   11855.0   48944.0   
29999         0.0         0.0  ...   36535.0   32428.0   15313.0   

       Montpago_Sep  Montpago_Ago  Montpago_Jul  Montpago_Jun  Montpago_May  \
0               0.0         689.0           0.0           0.0           0.0   
1               0.0        1000.0        1000.0        1000.0           0.0   
2            1518.0        1500.0        1000.0        1000.0        1000.0   
3            2000.0        2019.0        1200.0        1100.0        1069.0   
4            2000.0       36681.0       10000.0        9000.0         689.0   
...             ...           ...           ...           ...           ...   
29995        8500.0       20000.0        5003.0        3047.0        5000.0   
29996        1837.0        3526.0        8998.0         129.0           0.0   
29997           0.0           0.0       22000.0        4200.0        2000.0   
29998       85900.0        3409.0        1178.0        1926.0       52964.0   
29999        2078.0        1800.0        1430.0        1000.0        1000.0   

       Montpago_Abr    Y  
0               0.0  1.0  
1            2000.0  1.0  
2            5000.0  0.0  
3            1000.0  0.0  
4             679.0  0.0  
...             ...  ...  
29995        1000.0  0.0  
29996           0.0  0.0  
29997        3100.0  1.0  
29998        1804.0  1.0  
29999        1000.0  1.0  

[29958 rows x 25 columns]

**Parte 3: Preparación de los datos**


1. ¿Qué datos considero mas importantes? ¿Por qué?
Dentro de los tabla de datos considero que varias columnas son importantes para el análisis, así como desde la edad o género solo para generar insights en cuales tiene más ponderación para clientes deudores o no deudores y así poder construir un modelo automático para futuros nuevos clientes y poder determinar que crédito se le puede aprobar como también las variables muy importantes que son de sus pagos mes con mes y su monto deudor total.

                
2. ¿Se eliminaron o reemplazaron datos nulos? ¿Qué se hizo y por qué?

Se eliminaron los datos nulos ya que solo representaban tan solo el .65% del total de la base de datos, determinando que esto no tiene un gran impacto al modelo que se construyó excluyendo estos. Así mismo se imputaron otros valores como la moda y media para Edad y Género para ver los resultados de estos, como conclusión después de esto creo que se pueden tener mejor resultados imputando valores pero para este caso en específico puede ser un cambio muy mínimo al eliminar valores nulos. 
                
3. ¿Es necesario limpiar los datos para el análisis? Sí / No / ¿Por qué?

Fue necesario renombrar las columnas y poder identificar fácilmente cada uno de ellas.

Si, y depende de cada base de datos, en específico de esta base fue necesario ya que varios valores pueden influir erróneos pueden influir en nuestros resultados y se puede desviar el resultado óptimo del modelo.
                
                
4. ¿Existen problemas de formato que deban solucionar antes del proceso de modelado? Sí / No / Por qué.

Sí, varias columnas tienen valores atípicos, como educación con valores fuera de la métrica 1-4 y en la base de datos se tienen valores 5-6, estos sin significar nada.
                
5. ¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?
Los datos de pagos, billing se dejaron igual ya que considero que esos datos no se deben modificar y son parte elemental para el análisis, 
